automation/prepare-and-start.sh

```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+set -e
+
+DOCKER_TIMEOUT=20 # Wait 20 seconds for docker to start
+
+cleanup() {
+    echo "[INFO] Running cleanup..."
+
+    # Stop docker gracefully
+    echo "[INFO] Stopping in container docker..."
+    DOCKERPIDFILE=/var/run/docker.pid
+    if [ -f $DOCKERPIDFILE ] && [ -s $DOCKERPIDFILE ] && ps $(cat $DOCKERPIDFILE) | grep -q docker; then
+        kill $(cat $DOCKERPIDFILE)
+        # Now wait for it to die
+        STARTTIME=$(date +%s)
+        ENDTIME=$(date +%s)
+        while [ -f $DOCKERPIDFILE ] && [ -s $DOCKERPIDFILE ] && ps $(cat $DOCKERPIDFILE) | grep -q docker; do
+            if [ $(($ENDTIME - $STARTTIME)) -le $DOCKER_TIMEOUT ]; then
+                sleep 1
+                ENDTIME=$(date +%s)
+            else
+                echo "[ERROR] Timeout while waiting for in container docker to die."
+                exit 1
+            fi
+        done
+    else
+        echo "[WARN] Can't stop docker container."
+        echo "[WARN] Your host might have been left with unreleased resources (ex. loop devices)."
+    fi
+
+    if [ "$1" == "fail" ]; then
+        exit 1
+    fi
+}
+trap 'cleanup fail' SIGINT SIGTERM
+
+# Create the normal user to be used for bitbake (barys)
+echo "[INFO] Creating and setting builder user $BUILDER_UID:$BUILDER_GID."
+groupadd docker
+groupadd -g $BUILDER_GID builder
+useradd -m -u $BUILDER_UID -g $BUILDER_GID -G docker builder
+
+# Start docker
+echo "[INFO] Starting docker."
+docker daemon 2> /dev/null &
+echo "[INFO] Waiting for docker to initialize..."
+STARTTIME=$(date +%s)
+ENDTIME=$(date +%s)
+until docker info >/dev/null 2>&1; do
+    if [ $(($ENDTIME - $STARTTIME)) -le $DOCKER_TIMEOUT ]; then
+        sleep 1
+        ENDTIME=$(date +%s)
```

Comment:
What if a SIGINT or SIGTERM arrives here? Then cleanup() would be invoked, and it would go through the

    else
 +        echo "[WARN] Can't stop docker container."
 +        echo "[WARN] Your host might have been left with unreleased resources (ex. loop devices)."
 +    fi

branch, without attempting to kill docker as it is not completely started yet. And it may start after this.
So what if you also pass cleanup() the amount of seconds left until DOCKER_TIMEOUT is reached so there is a chance for docker to start correctly and then to actually try to terminate it so that docker can free up his resources correctly?